File: apps/web/lib/selection_range.ts
/** 触控词选范围（支持跨节） */

export type WordAnchor = { verse: number; start: number; end: number };

export type WordRange = { anchor: WordAnchor; focus: WordAnchor };

export function normalizeWordRange(range: WordRange): {
  verses: number[];
  anchor: WordAnchor;
  focus: WordAnchor;
} {
  const a = range.anchor;
  const f = range.focus;
  if (a.verse < f.verse || (a.verse === f.verse && a.start <= f.start)) {
    const verses = Array.from({ length: f.verse - a.verse + 1 }, (_, i) => a.verse + i);
    return { verses, anchor: a, focus: f };
  }
  const verses = Array.from({ length: a.verse - f.verse + 1 }, (_, i) => f.verse + i);
  return { verses, anchor: f, focus: a };
}

export function wordRangeToSpan(
  range: WordRange,
): { verses: number[]; span: { start: number; end: number } | null } {
  const { verses, anchor, focus } = normalizeWordRange(range);
  if (verses.length === 1) {
    const lo = Math.min(anchor.start, focus.start);
    const hi = Math.max(anchor.end, focus.end);
    return { verses, span: hi > lo ? { start: lo, end: hi } : null };
  }
  return { verses, span: null };
}

export function textFromWordRange(
  range: WordRange,
  verseText: (verse: number) => string,
): string {
  const { verses, anchor, focus } = normalizeWordRange(range);
  const loV = anchor.verse;
  const hiV = focus.verse;
  let out = '';
  for (const v of verses) {
    const text = verseText(v);
    if (loV === hiV) {
      out += text.slice(Math.min(anchor.start, focus.start), Math.max(anchor.end, focus.end));
    } else if (v === loV) {
      out += text.slice(anchor.start);
    } else if (v === hiV) {
      out += text.slice(0, focus.end);
    } else {
      out += text;
    }
  }
  return out;
}

export function wordOverlapsRange(
  verse: number,
  wordStart: number,
  wordEnd: number,
  range: WordRange,
): boolean {
  const { verses, anchor, focus } = normalizeWordRange(range);
  if (!verses.includes(verse)) return false;
  const loV = anchor.verse;
  const hiV = focus.verse;
  if (loV === hiV) {
    const lo = Math.min(anchor.start, focus.start);
    const hi = Math.max(anchor.end, focus.end);
    return wordStart < hi && wordEnd > lo;
  }
  if (verse === loV) return wordEnd > anchor.start;
  if (verse === hiV) return wordStart < focus.end;
  return true;
}

File: apps/mobile/lib/features/bible/selection_range.dart
/// 触控词选范围（对齐 Web `selection_range.ts`）。
library;

class WordAnchor {
  const WordAnchor({
    required this.verse,
    required this.start,
    required this.end,
  });
  final int verse;
  final int start;
  final int end;
}

class WordRange {
  const WordRange({required this.anchor, required this.focus});
  final WordAnchor anchor;
  final WordAnchor focus;
}

({List<int> verses, WordAnchor anchor, WordAnchor focus}) normalizeWordRange(
  WordRange range,
) {
  final a = range.anchor;
  final f = range.focus;
  if (a.verse < f.verse || (a.verse == f.verse && a.start <= f.start)) {
    final verses = [for (var i = a.verse; i <= f.verse; i++) i];
    return (verses: verses, anchor: a, focus: f);
  }
  final verses = [for (var i = f.verse; i <= a.verse; i++) i];
  return (verses: verses, anchor: f, focus: a);
}

({List<int> verses, ({int start, int end})? span}) wordRangeToSpan(
  WordRange range,
) {
  final n = normalizeWordRange(range);
  if (n.verses.length == 1) {
    final lo = n.anchor.start < n.focus.start ? n.anchor.start : n.focus.start;
    final hi = n.anchor.end > n.focus.end ? n.anchor.end : n.focus.end;
    return (
      verses: n.verses,
      span: hi > lo ? (start: lo, end: hi) : null,
    );
  }
  return (verses: n.verses, span: null);
}

String textFromWordRange(
  WordRange range,
  String Function(int verse) verseText,
) {
  final n = normalizeWordRange(range);
  final loV = n.anchor.verse;
  final hiV = n.focus.verse;
  final buf = StringBuffer();
  for (final v in n.verses) {
    final text = verseText(v);
    if (loV == hiV) {
      final lo = n.anchor.start < n.focus.start ? n.anchor.start : n.focus.start;
      final hi = n.anchor.end > n.focus.end ? n.anchor.end : n.focus.end;
      buf.write(text.substring(lo.clamp(0, text.length), hi.clamp(0, text.length)));
    } else if (v == loV) {
      buf.write(text.substring(n.anchor.start.clamp(0, text.length)));
    } else if (v == hiV) {
      buf.write(text.substring(0, n.focus.end.clamp(0, text.length)));
    } else {
      buf.write(text);
    }
  }
  return buf.toString();
}

bool wordOverlapsRange(
  int verse,
  int wordStart,
  int wordEnd,
  WordRange range,
) {
  final n = normalizeWordRange(range);
  if (!n.verses.contains(verse)) return false;
  final loV = n.anchor.verse;
  final hiV = n.focus.verse;
  if (loV == hiV) {
    final lo = n.anchor.start < n.focus.start ? n.anchor.start : n.focus.start;
    final hi = n.anchor.end > n.focus.end ? n.anchor.end : n.focus.end;
    return wordStart < hi && wordEnd > lo;
  }
  if (verse == loV) return wordEnd > n.anchor.start;
  if (verse == hiV) return wordStart < n.focus.end;
  return true;
}

WordRange wholeVerseRange(int verse, String text) => WordRange(
      anchor: WordAnchor(verse: verse, start: 0, end: text.length),
      focus: WordAnchor(verse: verse, start: 0, end: text.length),
    );
